### PR TITLE
ASoC: SOF: ipc4: get pipeline instance id from pipe_widget->instance_id

### PR DESCRIPTION
**Pull request by bardliao**
Currently we set pipeline instance id = swidget->pipeline_id, but
pipeline_id is from topology and can be any number. In fact, the
number of pipelines is limited. This patch use ida to allocate pipeline
instance id and will be used for pipeline instance id in the IPC
message.

### CODE REVIEW

File: sound/soc/sof/ipc4-loader.c [hunk 1]
```diff
@@ -120,6 +121,14 @@ static size_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev)
 		fm_entry++;
 	}
 
+	/* Add pipeline fw_module */
+	fw_module->man4_module_entry.id = i;
+	fw_module->man4_module_entry.uuid = SOF_IPC4_PIPELINE_MODULE_GUID;
+	fw_module->man4_module_entry.instance_max_count = SOF_IPC4_PPL_NUM;
+	strcpy(fw_module->man4_module_entry.name, "PIPELINE");
+	ida_init(&fw_module->m_ida);
+	ipc4_data->num_fw_modules++;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-loader.c]: Sorry, I am completely missing the point here. pipelines and modules are different things. That doesn't seem very elegant to me to conflate concepts.

Besides, what if we have more than 24 modules, wouldn't we hit the same issue in the IPC?

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -280,7 +280,7 @@ static void sof_ipc4_free_audio_fmt(struct sof_ipc4_available_audio_format *avai
 	available_fmt->out_audio_fmt = NULL;
 }
 
-static void sof_ipc4_widget_free_comp(struct snd_sof_widget *swidget)
+static void sof_ipc4_widget_free_comp_ppl(struct snd_sof_widget *swidget)
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: sof_ipc4_scheduler_widget_free()?

File: sound/soc/sof/ipc4-loader.c [hunk 3]
```diff
@@ -66,8 +66,9 @@ static size_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev)
 	dev_dbg(sdev->dev, "Firmware name: %s, header length: %u, module count: %u\n",
 		fw_header->name, fw_header->len, fw_header->num_module_entries);
 
+	/* allocate one more fw_module for pipeline */
```
Comment (ranj063) [inline sound/soc/sof/ipc4-loader.c]: @bardliao I think this step is unnecessary. You can add a case SOF_IPC4_FW_CFG_MAX_PPL_COUNT to sof_ipc4_query_fw_configuration() to query the max pipeline count supported by the FW and simply use that instead of adding a new FW module.

**Comment by bardliao** [comment]
@ranj063 @plbossart Thanks for your suggestion. I don't create a new module for pipeline now. I get max pipeline number as @ranj063 suggested and allocate the ida when a pcm is started. So it is fine if we use less pipelines than the firmware can support at the same time.

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > sof_ipc4_scheduler_widget_free()?

This is followed the naming rule of other widgets. Like `sof_ipc4_widget_free_comp_pcm`, `sof_ipc4_widget_free_comp_dai`, `sof_ipc4_widget_free_comp_pga` etc

File: sound/soc/sof/ipc4-topology.c [hunk 4]
```diff
@@ -1462,6 +1465,15 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
...
 		msg->primary |= pipeline->mem_usage;
+
+		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->num_ppl, GFP_KERNEL);
+		if (swidget->instance_id < 0) {
+			dev_err(sdev->dev, "failed to assign instance id for %s: %d\n",
+				swidget->widget->name, swidget->instance_id);
+			return swidget->instance_id;
+		}
+		msg->primary &= ~SOF_IPC4_GLB_PIPE_INSTANCE_MASK;
+		msg->primary |= SOF_IPC4_GLB_PIPE_INSTANCE_ID(swidget->instance_id);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1464]: I am still not clear on the concepts
pipeline_id comes from topology, it will be used to generate all the control names
instance_id is allocated with an IDA

How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we *require* that the number of pipelines is compatible with firmware requirements.

Is this to promote topology reuse, e.g. in some cases we not include parts of the topology but we want to keep topology-defined pipeline_id fixed?

It would help to better explain the problem statement and goals.

File: sound/soc/sof/ipc4-topology.c [hunk 5]
```diff
@@ -1535,7 +1547,11 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg->extension &= ~SOF_IPC4_MOD_EXT_PARAM_SIZE_MASK;
 		msg->extension |= ipc_size >> 2;
+
+		msg->extension &= ~SOF_IPC4_MOD_EXT_PPL_ID_MASK;
+		msg->extension |= SOF_IPC4_MOD_EXT_PPL_ID(pipe_widget->instance_id);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1540]: I am confused about this. Isn't the need to set the PPL_ID information orthogonal with how the pipeline numbering is done? Shouldn't this be a separate patch to improve/harden the existing IPC4?

**Review by plbossart — COMMENTED** [review]
somewhat lost @bardliao , the code looks good but the problem isn't clear to me.

File: sound/soc/sof/ipc4-loader.c [hunk 6]
```diff
@@ -193,6 +193,10 @@ static int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 			trace_sof_ipc4_fw_config(sdev, "Trace log size", *tuple->value);
 			ipc4_data->mtrace_log_bytes = *tuple->value;
 			break;
+		case SOF_IPC4_FW_CFG_MAX_PPL_COUNT:
+			ipc4_data->num_ppl = *tuple->value;
+			trace_sof_ipc4_fw_config(sdev, "Max PPL count %d", ipc4_data->num_ppl);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-loader.c]: Could you please split this part into a separate patch first?

File: sound/soc/sof/ipc4-priv.h [hunk 7]
```diff
@@ -40,6 +40,7 @@ struct sof_ipc4_fw_data {
 	void *nhlt;
 	enum sof_ipc4_mtrace_type mtrace_type;
 	u32 mtrace_log_bytes;
+	int num_ppl;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-priv.h]: also add to the description in the comment above? and maybe rename this to max_num_pipelines?

File: sound/soc/sof/ipc4-topology.c [hunk 8]
```diff
@@ -636,7 +637,7 @@ static int sof_ipc4_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
 	swidget->private = pipeline;
 
 	pipeline->msg.primary = SOF_IPC4_GLB_PIPE_PRIORITY(pipeline->priority);
-	pipeline->msg.primary |= SOF_IPC4_GLB_PIPE_INSTANCE_ID(swidget->pipeline_id);
+	pipeline->msg.primary |= SOF_IPC4_GLB_PIPE_INSTANCE_ID(swidget->instance_id);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: this should be removed isnt it? you dont really assign the pipeline ID until the widget is actually set up

File: sound/soc/sof/ipc4-topology.c [hunk 9]
```diff
@@ -1535,7 +1547,11 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg->extension &= ~SOF_IPC4_MOD_EXT_PARAM_SIZE_MASK;
 		msg->extension |= ipc_size >> 2;
+
+		msg->extension &= ~SOF_IPC4_MOD_EXT_PPL_ID_MASK;
+		msg->extension |= SOF_IPC4_MOD_EXT_PPL_ID(pipe_widget->instance_id);
 	}
+
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: spurious change?

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1540]: @bardliao you need to remove the pipeline ID seting on line 332 and like Pierre suggested do this in a separate patch

File: sound/soc/sof/ipc4-topology.c (hunk 4 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1464]: > I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> 
> How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.

The total number of pipeline may exceed the max number of pipeline that the FW can support.
The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16.
And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines.
The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.

> 
> Is this to promote topology reuse, e.g. in some cases we not include parts of the topology but we want to keep topology-defined pipeline_id fixed?
> 
> It would help to better explain the problem statement and goals.

Besides, assign instance_id in kernel giving more flexibility to topology.

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1540]: > @bardliao you need to remove the pipeline ID seting on line 332 and like Pierre suggested do this in a separate patch

done

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1540]: @plbossart @ranj063 I don't think it belongs to a separate patch. The existing code use a fixed pipeline id which is assigned when loading topology and will not be changed. The patch needs this because we assign instance_id dynamically.

File: sound/soc/sof/ipc4-topology.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: this isnt removed still

File: sound/soc/sof/ipc4-topology.c [hunk 10]
```diff
@@ -1462,6 +1464,15 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
+		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->num_ppl, GFP_KERNEL);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: does this compile? i thought you renamed it to max_num _pipelines?

File: sound/soc/sof/ipc4-topology.c [hunk 11]
```diff
@@ -1462,6 +1464,15 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
...
+		if (swidget->instance_id < 0) {
+			dev_err(sdev->dev, "failed to assign instance id for %s: %d\n",
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: can we explicitly says pipeline ID instead of insteance ID?

File: sound/soc/sof/ipc4-topology.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > does this compile? i thought you renamed it to max_num _pipelines?

oops, fixed it now.

File: sound/soc/sof/ipc4-topology.c (hunk 9 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > this isnt removed still

Removed

File: sound/soc/sof/ipc4-topology.c [hunk 12]
```diff
@@ -1462,6 +1463,16 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.c:1455]: how about a pcm stream composed of many pipelines ? there will be only one scheduler so only need to allocate one ? how about others ?

File: sound/soc/sof/ipc4-topology.c (hunk 11 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > can we explicitly says pipeline ID instead of insteance ID?

ack

File: sound/soc/sof/ipc4-topology.c (hunk 12 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1455]: > how about a pcm stream composed of many pipelines ? there will be only one scheduler so only need to allocate one ? how about others ?

`sof_ipc4_widget_setup` will be triggered for each pipeline not just one scheduler. For example, we have two pipelines for SSP playback one for FE and the other for BE, `sof_ipc4_widget_setup` will be triggered for both pipelines and different pipeline (instance) id will be assigned.

File: sound/soc/sof/ipc4-topology.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:1464]: > > I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> > How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.
> 
> The total number of pipeline may exceed the max number of pipeline that the FW can support. The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16. And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines. The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.

I don't get it @bardliao 

Are you saying that the IDA is used when the pipeline is instantiated, so depending on how the PCMs are opened we may or may not exceed the limitations on the number of pipelines?

This would be a change from all previous topologies where the assumption was that all resources can be used concurrently. I have no idea how UCM/PipeWire/PulseAudio would behave if *sometimes* a pipeline cannot be opened. 

Actually related question: is the IDA allocation handled in open or hw_params?

**Review by ranj063 — APPROVED** [review]

File: sound/soc/sof/ipc4-topology.c (hunk 12 above)
Comment (RanderWang) [reply sound/soc/sof/ipc4-topology.c:1455]: but the allocation code is under "case snd_soc_dapm_scheduler:" ?

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1455]: @RanderWang widgets in a pipeline are setup only after the pipeline widget for the pipeline they belong to are set up. So this code is fine.

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:1540]: I don't get this either. Even if we used a fixed setting, why didn't we need the message extension before.

My point is that regardless of what value is used for the pipeline, the sequence is new, and I fail to see why it was not needed before.

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1540]: @plbossart we did use the message extension. It was done in sof_ipc4_widget_setup_msg() which is called before the pipeline widget is set up in the firmware with the pipeline ID that was statically allocated in topology. But now since it is allocated dynamically, this has been moved to  sof_ipc4_widget_setup() which comes after the pipeline widget is set up and the dynamic pipeline ID is available.

Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:1540]: @ranj063 I only see an addition in this patch, not a code move. What am I missing?

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1540]: @plbossart please see the removal of line 331/332 above

File: sound/soc/sof/ipc4-topology.c (hunk 4 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1464]: > > > I am still not clear on the concepts pipeline_id comes from topology, it will be used to generate all the control names instance_id is allocated with an IDA
> > > How does this help? we still need to make sure the pipeline_id is unique in the topology, so why can't we _require_ that the number of pipelines is compatible with firmware requirements.
> > 
> > 
> > The total number of pipeline may exceed the max number of pipeline that the FW can support. The max pipeline number is less than we thought. For example, I get the max pipeline number of TGL Up Xtreme board is 16. And We have more than one pipelines in a PCM. If we have 2 SDW, 1 DMIC and 4 HDMI PCMs, we may need more than 16 pipelines. The proposal is to assign instance_id in runtime. So 16 pipelines is likely enough.
> 
> I don't get it @bardliao
> 
> Are you saying that the IDA is used when the pipeline is instantiated, so depending on how the PCMs are opened we may or may not exceed the limitations on the number of pipelines?

Depends on how many PCMs are opened, yes.

> 
> This would be a change from all previous topologies where the assumption was that all resources can be used concurrently. I have no idea how UCM/PipeWire/PulseAudio would behave if _sometimes_ a pipeline cannot be opened.
> 

Ah, I didn't know the assumption. But what if the firmware can only support less pipelines then we need?
Say we have 1 headset, 1 speaker with echo reference, 1 DMIC, and 4 HDMI.
So we have 6 playbacks and 3 captures.
Currently we need 2 pipelines for each playback or capture. So we need 18 pipelines in the above case.
And TGL FW can only support 16 pipelines.

> Actually related question: is the IDA allocation handled in open or hw_params?

In open

**Review by RanderWang — DISMISSED** [review]
LGTM

**Comment by plbossart** [comment]
@bardliao can you rebase and resubmit to solve the conflicts? Thanks!

**Comment by bardliao** [comment]
> @bardliao can you rebase and resubmit to solve the conflicts? Thanks!

Yes, done

File: sound/soc/sof/ipc4-loader.c [hunk 13]
```diff
@@ -383,6 +383,11 @@ int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 			if (!ipc4_data->max_libs_count)
 				ipc4_data->max_libs_count = 1;
 			break;
+		case SOF_IPC4_FW_CFG_MAX_PPL_COUNT:
+			ipc4_data->max_num_pipelines = *tuple->value;
+			trace_sof_ipc4_fw_config(sdev, "Max PPL count %d",
+						 ipc4_data->max_num_pipelines);
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-loader.c:389]: check max_num_pipelines to avoid unexpected behavior.

Comment (bardliao) [reply sound/soc/sof/ipc4-loader.c:389]: > check max_num_pipelines to avoid unexpected behavior.

Thanks @RanderWang I added a test before
```
		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->max_num_pipelines,
						     GFP_KERNEL);
```

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/ipc4-topology.c [hunk 14]
```diff
@@ -2024,7 +2024,8 @@ static const struct sof_ipc_tplg_widget_ops tplg_ipc4_widget_ops[SND_SOC_DAPM_TY
 				  dai_token_list, ARRAY_SIZE(dai_token_list), NULL,
 				  sof_ipc4_prepare_copier_module,
 				  sof_ipc4_unprepare_copier_module},
-	[snd_soc_dapm_scheduler] = {sof_ipc4_widget_setup_comp_pipeline, sof_ipc4_widget_free_comp,
+	[snd_soc_dapm_scheduler] = {sof_ipc4_widget_setup_comp_pipeline,
+				    sof_ipc4_widget_free_comp_ppl,
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: @bardliao, can we not use `ppl` the first thing to come to mind is `people`... In other callbacks we are able to spell it out, we should do that here as well.

File: sound/soc/sof/ipc4-topology.c [hunk 15]
```diff
@@ -1451,6 +1452,21 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
+		if (ipc4_data->max_num_pipelines <= 0) {
+			dev_err(sdev->dev, "Invalid max_num_pipelines %d\n",
+				ipc4_data->max_num_pipelines);
+			return ipc4_data->max_num_pipelines;
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: If `ipc4_data->max_num_pipelines` is 0, then you return here with success?

File: sound/soc/sof/ipc4-topology.c [hunk 16]
```diff
@@ -1451,6 +1452,21 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
...
+		}
+		swidget->instance_id = ida_alloc_max(&ppl_ida, ipc4_data->max_num_pipelines,
+						     GFP_KERNEL);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c:1457]: XArray is the preferred over the deprecated IDA...

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:1457]: You might want to free the allocation in the error cases after this point.

File: sound/soc/sof/ipc4-topology.c [hunk 17]
```diff
@@ -20,6 +20,7 @@
 #define SOF_IPC4_TPLG_ABI_SIZE 6
 
 static DEFINE_IDA(alh_group_ida);
+static DEFINE_IDA(ppl_ida);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: `people`_ida? can we spell it out, please?

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: sound/soc/sof/ipc4-loader.c (hunk 13 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c:389]: I would have verified it here and fail early. You can not allocate anyways if `ipc4_data->max_num_pipelines <= 0`

File: sound/soc/sof/ipc4-topology.c (hunk 15 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > If `ipc4_data->max_num_pipelines` is 0, then you return here with success?

Moved the test to where we get `ipc4_data->max_num_pipelines` and return -EINVAL in error.

File: sound/soc/sof/ipc4-topology.c (hunk 16 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:1457]: Thanks @ujfalusi I added `ida_free(&pipeline_ida, swidget->instance_id);` in error. But, could you describe the XArray comment in more detail?  Which part do you prefer using XArray? Thanks

**Review by RanderWang — APPROVED** [review]
LGTM

File: sound/soc/sof/ipc4-topology.c (hunk 16 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:1457]: is this correct @ujfalusi that ida_ is deprecated? That's is true of *idr*, 

https://docs.kernel.org/next/core-api/idr.html?highlight=idr#idr-usage

````
The IDR interface is deprecated; please use the [XArray](https://docs.kernel.org/next/core-api/xarray.html) instead.
````
but that's a different API.

What makes you think the IDA is deprecated? We clearly just added it for all kinds of things recently.

**Review by ranj063 — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
need @ujfalusi clarifications on IDA objection.

File: sound/soc/sof/ipc4-topology.c [hunk 18]
```diff
@@ -20,6 +20,7 @@
 #define SOF_IPC4_TPLG_ABI_SIZE 6
 
 static DEFINE_IDA(alh_group_ida);
+static DEFINE_IDA(pipeline_ida);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c:23]: As it was pointed out by @plbossart on the library loading PR, the IDA is deprecated and XArray should be used, do we want to do that or do a conversion later?

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:23]: Never mind!

File: sound/soc/sof/ipc4-topology.c (hunk 16 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:1457]: True, sorry

**Review by ujfalusi — APPROVED** [review]
